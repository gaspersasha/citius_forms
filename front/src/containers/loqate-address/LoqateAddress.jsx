import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { InputText, CustomLink, Spinner } from '~components';
import { INPUT_STATUS } from '~constants';
import { endpointFetch } from '~utils';
import { Postcode } from '.';
import s from './styles/loqate-address.module.sass';

class LoqateAddress extends Component {
  static propTypes = {
    selectHandler: PropTypes.func.isRequired,
    formName: PropTypes.string.isRequired,
    subTitle: PropTypes.string,
    errorMessage: PropTypes.string,
    manualFields: PropTypes.arrayOf(
      PropTypes.shape({
        label: PropTypes.string.isRequired,
        name: PropTypes.string.isRequired,
        value: PropTypes.string.isRequired,
        status: PropTypes.string.isRequired,
        onChange: PropTypes.func.isRequired,
        required: PropTypes.bool,
        warning: PropTypes.string,
        id: PropTypes.string,
      })
    ).isRequired,
  };

  static defaultProps = {
    subTitle: '',
    errorMessage: '',
  };

  state = {
    isCheckingApi: false,
    isManual: false,
    isApiHaveError: false,
  };

  componentDidMount() {
    this.preCheckLoqateFindApi();
  }

  loqateHealthCheck =
    (type, fn = this.loqateApiError) =>
    (response) => {
      if (!response) {
        return fn();
      }

      if (!Object.prototype.hasOwnProperty.call(response, 'Items')) {
        return fn();
      }

      if (!response.Items.length) {
        return response;
      }

      const firstItem = response.Items[0];

      // Check Error item in response
      if (firstItem.Error && !firstItem.Cause && !firstItem.Resolution) {
        return fn();
      }

      // Check success response from Loqate API
      if (!firstItem.Error) {
        const invalidItem = response.Items.find(this.invalidResponse(type));

        if (invalidItem) {
          return fn();
        }
      }

      return response;
    };

  preCheckLoqateFindApi = () =>
    this.setState({ isCheckingApi: true }, () =>
      endpointFetch('loqateAPIFind', { query: 'Buy A Car' }, {}, false)
        .then(this.loqateHealthCheck('Find', this.switchAddressField))
        .then(this.turnOffLoader)
        .catch(console.warn)
    );

  turnOffLoader = () => this.setState({ isCheckingApi: false });

  switchAddressField = () =>
    this.setState(({ isManual }) => ({ isManual: !isManual }));

  formatAddress = (fields) =>
    fields.reduce(
      (prev, field) => ({ ...prev, [field.name]: field.value }),
      {}
    );

  getRequiredFields = () => {
    let postCode;
    let street;
    let town;

    this.props.manualFields.forEach((item) => {
      if (item.name === 'postCode') postCode = item;
      if (item.name === 'street') street = item;
      if (item.name === 'town') town = item;
    });

    return { postCode, street, town };
  };

  generateManualInputs = () =>
    this.props.manualFields.map(({ warning, ...props }) => (
      <div key={props.name} className="form-input">
        <InputText {...props} />
        {warning && props.status === INPUT_STATUS.INVALID && (
          <p data-id="manual-field-warn" className={s.warning}>
            {warning}
          </p>
        )}
      </div>
    ));

  loqateApiError = () => {
    this.setState({ isApiHaveError: true }, () => this.switchAddressField());

    return {};
  };

  invalidResponse = (type) => (item) => {
    const keys = Object.keys(item);

    switch (type) {
      case 'Retrieve':
        return (
          !keys.includes('Language') ||
          !keys.includes('City') ||
          !keys.includes('District') ||
          !keys.includes('PostalCode') ||
          !keys.includes('Street') ||
          !keys.includes('BuildingNumber') ||
          !keys.includes('SubBuilding') ||
          !keys.includes('BuildingName')
        );
      case 'Find':
      default:
        return (
          !item.Id ||
          !keys.includes('Type') ||
          !keys.includes('Text') ||
          !keys.includes('Description')
        );
    }
  };

  render() {
    const { formName, errorMessage } = this.props;
    const { isManual, isApiHaveError, isCheckingApi } = this.state;

    if (isCheckingApi) {
      return <Spinner />;
    }

    const selectedAddress = this.formatAddress(this.props.manualFields);
    const { postCode, street, town } = this.getRequiredFields();

    return (
      <>
        {isApiHaveError && (
          <div className="form_input_warning">
            Oops! Our address lookup is not responding. Please enter your
            address manually
          </div>
        )}

        {isManual ? (
          this.generateManualInputs()
        ) : (
          <div className="form-input">
            <Postcode
              formName={formName}
              subTitle={this.props.subTitle}
              styleType="loqate-address"
              errorMessage={errorMessage}
              selectHandler={this.props.selectHandler}
              statuses={{
                postCode: postCode.status,
                town: town.status,
                street: street.status,
              }}
              selectedAddress={selectedAddress}
              loqateHealthCheck={this.loqateHealthCheck}
            />
            {!postCode.value && (
              <div id={`${formName}ManuallyLink`} className={s.link}>
                <CustomLink
                  type="loqate-address-link__text"
                  onClick={this.switchAddressField}
                  data-id="enter-address-manually"
                >
                  Enter address manually
                </CustomLink>
              </div>
            )}
          </div>
        )}
      </>
    );
  }
}

export default LoqateAddress;
