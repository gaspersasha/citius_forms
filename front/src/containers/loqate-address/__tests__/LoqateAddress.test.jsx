import React from 'react';
import { INPUT_STATUS } from '~constants';
import { endpointFetch } from '~utils';
import LoqateAddress from '../LoqateAddress';

jest.mock('~utils', () => ({
  ...jest.requireActual('../../../utils'),
  endpointFetch: jest.fn(),
}));

const defaultProps = {
  selectHandler: jest.fn(),
  formName: '',
  manualFields: [
    {
      label: 'postCode label',
      name: 'postCode',
      value: 'test postCode',
      status: INPUT_STATUS.VALID,
      validationType: '',
      onChange: jest.fn(),
    },
    {
      label: 'town label',
      name: 'town',
      value: 'test town',
      status: INPUT_STATUS.VALID,
      validationType: '',
      onChange: jest.fn(),
    },
    {
      label: 'street label',
      name: 'street',
      value: 'test street',
      status: INPUT_STATUS.INVALID,
      validationType: '',
      onChange: jest.fn(),
      warning: 'Warning',
    },
  ],
};

describe('<LoqateAddress />', () => {
  let wrapper;

  const getWrapper = (props) =>
    shallow(<LoqateAddress {...defaultProps} {...props} />);

  it('renders without errors', async () => {
    endpointFetch.mockResolvedValueOnce(null);
    wrapper = getWrapper();

    await endpointFetch;
    await endpointFetch;
    wrapper.update();

    expect(wrapper).toMatchSnapshot();
  });

  it('unmounts without errors', () => {
    endpointFetch.mockResolvedValueOnce(null);
    wrapper = getWrapper();
    wrapper.unmount();
  });

  it('handles manual fields case (no response)', async () => {
    endpointFetch.mockResolvedValueOnce(null);
    wrapper = getWrapper();

    expect(wrapper.find('Spinner').exists()).toBe(true);

    await endpointFetch;
    await endpointFetch;
    wrapper.update();

    expect(wrapper.find('Spinner').exists()).toBe(false);

    expect(wrapper.find('InputText').at(0).props()).toEqual(
      expect.objectContaining(defaultProps.manualFields[0])
    );
    expect(wrapper.find('InputText').at(1).props()).toEqual(
      expect.objectContaining(defaultProps.manualFields[1])
    );

    expect(wrapper.find('[data-id="manual-field-warn"]').text()).toEqual(
      defaultProps.manualFields[2].warning
    );

    delete defaultProps.manualFields[2].warning;

    expect(wrapper.find('InputText').at(2).props()).toEqual(
      expect.objectContaining(defaultProps.manualFields[2])
    );
  });

  it('handles manual fields case (response without items)', async () => {
    endpointFetch.mockResolvedValueOnce({});
    wrapper = getWrapper();

    await endpointFetch;
    await endpointFetch;
    wrapper.update();

    expect(wrapper.find('InputText').at(0).props()).toEqual(
      expect.objectContaining(defaultProps.manualFields[0])
    );
    expect(wrapper.find('InputText').at(1).props()).toEqual(
      expect.objectContaining(defaultProps.manualFields[1])
    );
    expect(wrapper.find('InputText').at(2).props()).toEqual(
      expect.objectContaining(defaultProps.manualFields[2])
    );
  });

  it('handles manual fields case (empty response)', async () => {
    endpointFetch.mockResolvedValueOnce({ Items: [] });
    wrapper = getWrapper();

    await endpointFetch;
    await endpointFetch;
    wrapper.update();

    expect(wrapper.find('InputText').exists()).toBe(false);
    expect(wrapper).toMatchSnapshot();
  });

  it('handles manual fields case (api error)', async () => {
    endpointFetch.mockResolvedValueOnce({ Items: [{ Error: true }] });
    wrapper = getWrapper();

    await endpointFetch;
    await endpointFetch;
    wrapper.update();

    expect(wrapper.find('InputText').at(0).props()).toEqual(
      expect.objectContaining(defaultProps.manualFields[0])
    );
    expect(wrapper.find('InputText').at(1).props()).toEqual(
      expect.objectContaining(defaultProps.manualFields[1])
    );
    expect(wrapper.find('InputText').at(2).props()).toEqual(
      expect.objectContaining(defaultProps.manualFields[2])
    );
  });

  it('handles manual fields case (invalid items)', async () => {
    endpointFetch.mockResolvedValueOnce({ Items: [{ Error: false }] });
    wrapper = getWrapper();

    await endpointFetch;
    await endpointFetch;
    wrapper.update();

    expect(wrapper.find('InputText').at(0).props()).toEqual(
      expect.objectContaining(defaultProps.manualFields[0])
    );
    expect(wrapper.find('InputText').at(1).props()).toEqual(
      expect.objectContaining(defaultProps.manualFields[1])
    );
    expect(wrapper.find('InputText').at(2).props()).toEqual(
      expect.objectContaining(defaultProps.manualFields[2])
    );
  });

  it('handles loqateHealthCheck call from Postcode', async () => {
    endpointFetch.mockResolvedValueOnce({ Items: [] });
    wrapper = getWrapper();

    await endpointFetch;
    await endpointFetch;
    wrapper.update();

    wrapper.find('Postcode').invoke('loqateHealthCheck')('Retrieve')();
    expect(wrapper.find('[className="form_input_warning"]').exists()).toBe(
      true
    );
  });
});
