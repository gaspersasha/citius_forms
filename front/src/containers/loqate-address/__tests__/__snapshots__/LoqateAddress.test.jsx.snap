// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoqateAddress /> handles manual fields case (empty response) 1`] = `
<Fragment>
  <div
    className="form-input"
  >
    <Postcode
      errorMessage=""
      formName=""
      loqateHealthCheck={[Function]}
      selectHandler={[MockFunction]}
      selectedAddress={
        Object {
          "postCode": "test postCode",
          "street": "test street",
          "town": "test town",
        }
      }
      statuses={
        Object {
          "postCode": "valid",
          "street": "invalid",
          "town": "valid",
        }
      }
      styleType="loqate-address"
      subTitle=""
    />
  </div>
</Fragment>
`;

exports[`<LoqateAddress /> renders without errors 1`] = `
<Fragment>
  <div
    className="form-input"
    key="postCode"
  >
    <InputText
      className=""
      customClasses=""
      errorMessage=""
      label="postCode label"
      name="postCode"
      onChange={[MockFunction]}
      placeholder=""
      required={false}
      status="valid"
      subTitle=""
      validationType=""
      value="test postCode"
    />
  </div>
  <div
    className="form-input"
    key="town"
  >
    <InputText
      className=""
      customClasses=""
      errorMessage=""
      label="town label"
      name="town"
      onChange={[MockFunction]}
      placeholder=""
      required={false}
      status="valid"
      subTitle=""
      validationType=""
      value="test town"
    />
  </div>
  <div
    className="form-input"
    key="street"
  >
    <InputText
      className=""
      customClasses=""
      errorMessage=""
      label="street label"
      name="street"
      onChange={[MockFunction]}
      placeholder=""
      required={false}
      status="invalid"
      subTitle=""
      validationType=""
      value="test street"
    />
    <p
      className="warning"
      data-id="manual-field-warn"
    >
      Warning
    </p>
  </div>
</Fragment>
`;
